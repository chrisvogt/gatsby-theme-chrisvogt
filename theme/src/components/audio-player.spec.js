import React from 'react'
import renderer, { act } from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import AudioPlayer from '../components/audio-player'

// Mock SoundCloud and Spotify to avoid iframe logic
jest.mock('../shortcodes/soundcloud', () => jest.fn(() => <div>MockSoundCloud</div>))
jest.mock('../shortcodes/spotify', () => jest.fn(() => <div>MockSpotify</div>))

const mockStore = configureStore([])

describe('AudioPlayer', () => {
  let store

  beforeEach(() => {
    store = mockStore({})
    store.dispatch = jest.fn()
    document.body.innerHTML = ''
  })

  it('renders and cleans up the portal container', async () => {
    let component
    await act(async () => {
      component = renderer.create(
        <Provider store={store}>
          <AudioPlayer soundcloudId='abc' isVisible={true} provider='soundcloud' />
        </Provider>
      )
    })

    expect(document.getElementById('audio-player-portal')).toBeTruthy()

    await act(async () => {
      component.unmount()
    })

    expect(document.getElementById('audio-player-portal')).toBeNull()
  })

  it('matches snapshot when visible with SoundCloud', async () => {
    // Only mock createPortal for this test
    const createPortalMock = jest.spyOn(require('react-dom'), 'createPortal').mockImplementation(node => node)

    let component
    await act(async () => {
      component = renderer.create(
        <Provider store={store}>
          <AudioPlayer soundcloudId='abc' isVisible={true} provider='soundcloud' />
        </Provider>
      )
    })

    expect(component.toJSON()).toMatchSnapshot()

    createPortalMock.mockRestore()
  })

  it('matches snapshot when visible with Spotify', async () => {
    const createPortalMock = jest.spyOn(require('react-dom'), 'createPortal').mockImplementation(node => node)

    let component
    await act(async () => {
      component = renderer.create(
        <Provider store={store}>
          <AudioPlayer spotifyURL='https://spotify.com/track/123' isVisible={true} provider='spotify' />
        </Provider>
      )
    })

    expect(component.toJSON()).toMatchSnapshot()

    createPortalMock.mockRestore()
  })

  it('does not render when not visible', async () => {
    let component
    await act(async () => {
      component = renderer.create(
        <Provider store={store}>
          <AudioPlayer soundcloudId='abc' isVisible={false} provider='soundcloud' />
        </Provider>
      )
    })

    expect(component.toJSON()).toBeNull()
  })

  it('does not render when no provider', async () => {
    let component
    await act(async () => {
      component = renderer.create(
        <Provider store={store}>
          <AudioPlayer soundcloudId='abc' isVisible={true} />
        </Provider>
      )
    })

    expect(component.toJSON()).toBeNull()
  })

  it('does not render when container ref is not available', async () => {
    // Mock createPortal to return null to simulate no container
    const createPortalMock = jest.spyOn(require('react-dom'), 'createPortal').mockImplementation(() => null)

    let component
    await act(async () => {
      component = renderer.create(
        <Provider store={store}>
          <AudioPlayer soundcloudId='abc' isVisible={true} provider='soundcloud' />
        </Provider>
      )
    })

    expect(component.toJSON()).toBeNull()

    createPortalMock.mockRestore()
  })

  it('renders SoundCloud when provider is soundcloud and soundcloudId is provided', async () => {
    // Test that the component doesn't throw when rendered with valid props
    await act(async () => {
      renderer.create(
        <Provider store={store}>
          <AudioPlayer soundcloudId='abc123' isVisible={true} provider='soundcloud' />
        </Provider>
      )
    })

    // If we get here without throwing, the test passes
    expect(true).toBe(true)
  })

  it('renders Spotify when provider is spotify and spotifyURL is provided', async () => {
    // Test that the component doesn't throw when rendered with valid props
    await act(async () => {
      renderer.create(
        <Provider store={store}>
          <AudioPlayer spotifyURL='https://spotify.com/track/123' isVisible={true} provider='spotify' />
        </Provider>
      )
    })

    // If we get here without throwing, the test passes
    expect(true).toBe(true)
  })

  it('renders null when provider is soundcloud but no soundcloudId', async () => {
    const createPortalMock = jest.spyOn(require('react-dom'), 'createPortal').mockImplementation(node => node)

    await act(async () => {
      renderer.create(
        <Provider store={store}>
          <AudioPlayer isVisible={true} provider='soundcloud' />
        </Provider>
      )
    })

    expect(require('../shortcodes/soundcloud')).not.toHaveBeenCalled()

    createPortalMock.mockRestore()
  })

  it('renders null when provider is spotify but no spotifyURL', async () => {
    const createPortalMock = jest.spyOn(require('react-dom'), 'createPortal').mockImplementation(node => node)

    await act(async () => {
      renderer.create(
        <Provider store={store}>
          <AudioPlayer isVisible={true} provider='spotify' />
        </Provider>
      )
    })

    expect(require('../shortcodes/spotify')).not.toHaveBeenCalled()

    createPortalMock.mockRestore()
  })

  it('renders null when provider is unknown', async () => {
    const createPortalMock = jest.spyOn(require('react-dom'), 'createPortal').mockImplementation(node => node)

    await act(async () => {
      renderer.create(
        <Provider store={store}>
          <AudioPlayer
            soundcloudId='abc'
            spotifyURL='https://spotify.com/track/123'
            isVisible={true}
            provider='unknown'
          />
        </Provider>
      )
    })

    expect(require('../shortcodes/soundcloud')).not.toHaveBeenCalled()
    expect(require('../shortcodes/spotify')).not.toHaveBeenCalled()

    createPortalMock.mockRestore()
  })

  it('dispatches hidePlayer when close button is clicked', async () => {
    const createPortalMock = jest.spyOn(require('react-dom'), 'createPortal').mockImplementation(node => node)

    await act(async () => {
      renderer.create(
        <Provider store={store}>
          <AudioPlayer soundcloudId='abc' isVisible={true} provider='soundcloud' />
        </Provider>
      )
    })

    // Test that the dispatch function is available
    expect(store.dispatch).toBeDefined()

    createPortalMock.mockRestore()
  })

  it('updates widget ref when soundcloudId changes', async () => {
    // Test that the component can be rendered with different soundcloudId values
    await act(async () => {
      renderer.create(
        <Provider store={store}>
          <AudioPlayer soundcloudId='abc' isVisible={true} provider='soundcloud' />
        </Provider>
      )
    })

    // If we get here without throwing, the test passes
    expect(true).toBe(true)
  })
})
