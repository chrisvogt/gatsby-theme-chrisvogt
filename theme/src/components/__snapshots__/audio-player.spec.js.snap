// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AudioPlayer matches snapshot when visible with SoundCloud 1`] = `null`;

exports[`AudioPlayer matches snapshot when visible with Spotify 1`] = `null`;
